src/Pos/Block/Logic.hs

```diff
@@ -147,6 +149,34 @@ lcaWithMainChain headers =
             ([], True)   -> pure $ Just h
             (x:xs, True) -> lcaProceed (Just h) (x :| xs)
 
+-- | The phrase “we're in recovery mode” is confusing because it can mean two
+-- different things:
+--
+-- 1. Last known block is more than K slots away from the current slot, or
+--    current slot isn't known.
+--
+-- 2. We're actually in the process of requesting blocks because we have
+--    detected that #1 happened. (See 'ncRecoveryHeader' and
+--    'recoveryInProgress'.)
+--
+-- This function checks for #1. Note that even if we're doing recovery right
+-- now, 'needRecovery' will still return 'True'.
+--
+needRecovery
+    :: forall ssc m.
+       (SscWorkersClass ssc, WorkMode ssc m)
+    => m Bool
+needRecovery =
+    getCurrentSlot >>= \case
+        Nothing   -> pure True
+        Just slot -> isTooOld slot
+  where
+    isTooOld slot = do
+        knownSlot <- getEpochOrSlot <$> DB.getTipBlockHeader @ssc
+        pure $
+            (knownSlot & from enum %~ (+ blkSecurityParam)) <
```

Comment:
> Also are you 100% sure converting to/from enum won't break the getEpochOrSlot?

Well, there is some semi-complicated instance for `Enum EpochOrSlot` (I think @gromakovsky wrote it), but I guess it's actually inappropriate in this case:

```
> fromEnum (EpochOrSlot (Left 1))
21
> fromEnum (EpochOrSlot (Right (SlotId 1 0)))
22
```

The “genesis slot” doesn't take any time so it shouldn't be counted in `isTooOld`. I'll fix it.